app/models/embeddable_media.rb

```diff
@@ -2,15 +2,146 @@
+        else
+          # There are a handful of migrations records missing the 'original_url' value
+          self.original_url = parsed[1]
+        end
+      end
+      # Capture and honor the cell range entered into the legacy system.
+      # Should perhaps confirm with Brendan if that is desired, given that this functionality was
+      # broken in the legacy system.
+      self.start_cell = parsed[2]
+      self.end_cell = parsed[3]
+      # Beware the gid appended to the tail of the legacy system url is usually bogus and irrelevant
+    end
+    parse_key_gid_from_original_url
+    self
+  end
+
+  def parse_key_gid_from_original_url
+    self.original_url ||= url  # There are a few records in legacy db without original_url
+
+    # Extract document id from either new or old style document urls
+    ccc_style = /.*spreadsheet\/ccc\?key=(\w*).*/.match(original_url)
+    d_style = /.*spreadsheets\/d\/([\w-]*).*/.match(original_url)
+    key_match = ccc_style || d_style
+    self.document_key = key_match[1] if key_match
+
+    # Extract the sheet number from the original url.
+    # Note, the attempt to allow a user enter a sheet number or name in the popup form of the
+    # legacy PHP system never worked.
+    # That value was sipmly ignored and the gid from the orignal url actually being used.
+    gid_match = /.*gid=(\d*).*/.match(original_url)
+    self.sheet_number = gid_match[1] if gid_match
+    self
+  end
+
+  # To be discussed:
+  #   Should we remove url, width & height from the database?
+
+  # Display url logic from legacy PHP system
+  # def display_url
+  #   "#{original_url}&single=true&range=#{start_cell}%3A#{end_cell}&output=html&gid=#{sheet}"
+  # end
+
+  def display_url
+    ensure_migration
+    # Todo: confirm if any special behavior needed for Google Apps environments
+    if /.*spreadsheet\/ccc\?.*/.match(original_url)
+      "https://docs.google.com/spreadsheet/ccc?key=#{document_key}&gid=#{sheet_number}&single=true"\
+        "#{range_param}&output=html"
+    elsif /.*spreadsheets\/d\/.*/.match(original_url)
+      "https://docs.google.com/spreadsheets/d/#{document_key}/htmlembed?single=true"\
+        "&gid=#{sheet_number}#{range_param}&widget=false"
+    else
+      # Unexpected url format, just pass through
+      original_url
+    end
+  end
+
+  # In the php system, there used to be some logic dependent upon the range column count, but then
+  # later hardcoded to 600.
+  def display_width
+    600
+  end
+
+  def display_height
+    ensure_migration
+    (end_row - start_row) * 20 + 100
+  end
+
+  def start_row
+    row_from_cell(start_cell)
+  end
+
+  def end_row
+    row_from_cell(end_cell)
+  end
+
+  private
+
+  def range_param
+    if start_cell.present? && end_cell.present?
+      "&range=#{start_cell}%3A#{end_cell}"
+    else
+      ""
+    end
+  end
+
+  def row_from_cell(cell)
+    parsed = /\D+(\d+)/.match(cell)
+    parsed ? parsed[1].to_i : 0
+  end
+
```

Comment:
Is EmbeddableMedia really a good name for this class now? It's really just about spreadsheets, right?